File: app/src/main/java/com/ilyap/yuta/ui/fragments/ProjectsFragment.java
package com.ilyap.yuta.ui.fragments;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.ToggleButton;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.ilyap.yuta.R;
import com.ilyap.yuta.models.ProjectDto;
import com.ilyap.yuta.models.ProjectsResponse;
import com.ilyap.yuta.ui.adapters.ProjectsAdapter;
import com.ilyap.yuta.ui.dialogs.CustomDialog;
import com.ilyap.yuta.ui.dialogs.project.CreateProjectDialog;
import com.ilyap.yuta.ui.dialogs.user.LogoutDialog;
import com.ilyap.yuta.utils.RequestViewModel;
import lombok.NoArgsConstructor;

import java.util.ArrayList;
import java.util.List;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static com.ilyap.yuta.ui.dialogs.project.CreateProjectDialog.PICK_PDF_REQUEST;
import static com.ilyap.yuta.utils.UserUtils.getUserId;

@NoArgsConstructor
public class ProjectsFragment extends Fragment {
    private static int lastPickedButtonId;
    public final ActivityResultLauncher<Intent> pdfPickerLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == Activity.RESULT_OK) {
                    CreateProjectDialog.handleActivityResult(PICK_PDF_REQUEST, Activity.RESULT_OK, result.getData());
                }
            }
    );
    private ToggleButton managedProjectsButton;
    private ToggleButton memberProjectsButton;
    private TextView emptyText;
    private View progressLayout;
    private View view;
    private RequestViewModel viewModel;
    private List<ProjectDto> managedProjectsMembers;
    private List<ProjectDto> othersProjectsMembers;
    private ProjectsAdapter projectsAdapter;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragment_projects, container, false);

        emptyText = view.findViewById(R.id.empty_text);
        progressLayout = view.findViewById(R.id.progressLayout);
        progressLayout.setVisibility(VISIBLE);

        viewModel = new ViewModelProvider(this).get(RequestViewModel.class);

        recyclerViewInitialize();
        projectsSwitchInitialize();

        if (lastPickedButtonId == 0) {
            lastPickedButtonId = memberProjectsButton.getId();
        }

        view.findViewById(R.id.create_project).setOnClickListener(v -> openCreateProjectDialog());
        view.findViewById(R.id.log_out).setOnClickListener(v -> openLogoutDialog());
        return view;
    }

    @Override
    public void onStart() {
        super.onStart();
        updateLists();
    }

    private void fillProjects(@NonNull List<ProjectDto> projectMembers) {
        emptyText.setVisibility(projectMembers.isEmpty() ? VISIBLE : GONE);
        projectsAdapter.updateList(projectMembers);
    }

    private void getProjects() {
        viewModel.getResultLiveData().removeObservers(getViewLifecycleOwner());
        viewModel.getProjects(getUserId(requireActivity()));
        viewModel.getResultLiveData().observe(getViewLifecycleOwner(), result -> {
            if (!(result instanceof ProjectsResponse)) return;
            progressLayout.setVisibility(GONE);
            ProjectsResponse projectsResponse = (ProjectsResponse) result;
            managedProjectsMembers = projectsResponse.getManagedProjects();
            othersProjectsMembers = projectsResponse.getOthersProjects();
        });
    }

    public void updateLists() {
        updateLists(view.findViewById(lastPickedButtonId));
    }

    private void updateLists(Button button) {
        getProjects();
        viewModel.getResultLiveData().observe(getViewLifecycleOwner(), result -> {
            if (!(result instanceof ProjectsResponse)) return;
            openTab(button);
        });
    }

    private void openTab(@NonNull Button button) {
        button.performClick();
    }

    private void recyclerViewInitialize() {
        RecyclerView recyclerView = view.findViewById(R.id.recyclerView);
        LinearLayoutManager layoutManager = new LinearLayoutManager(requireContext(), LinearLayoutManager.VERTICAL, false);
        recyclerView.setLayoutManager(layoutManager);

        List<ProjectDto> projectList = new ArrayList<>();
        projectsAdapter = new ProjectsAdapter(requireActivity(), projectList, this);
        recyclerView.setAdapter(projectsAdapter);
    }

    private void onToggleButtonClick(View view) {
        final ToggleButton button = (ToggleButton) view;
        final ToggleButton otherButton;

        if (button.getId() == managedProjectsButton.getId()) {
            otherButton = memberProjectsButton;
            fillProjects(managedProjectsMembers);
        } else {
            otherButton = managedProjectsButton;
            fillProjects(othersProjectsMembers);
        }

        button.setTextAppearance(R.style.active_toggle);
        button.setChecked(true);
        otherButton.setTextAppearance(R.style.default_toggle);
        otherButton.setChecked(false);
        lastPickedButtonId = button.getId();
    }

    private void projectsSwitchInitialize() {
        managedProjectsButton = view.findViewById(R.id.manager_button);
        memberProjectsButton = view.findViewById(R.id.member_button);
        managedProjectsButton.setOnClickListener(this::onToggleButtonClick);
        memberProjectsButton.setOnClickListener(this::onToggleButtonClick);
    }

    private void openCreateProjectDialog() {
        CustomDialog createProjectDialog = new CreateProjectDialog(view.getContext(), this);
        createProjectDialog.start();
    }

    private void openLogoutDialog() {
        CustomDialog logoutDialog = new LogoutDialog(view.getContext(), this);
        logoutDialog.start();
    }
}

File: app/src/main/java/com/yuta/projects/ui/ProjectsFragment.kt
package com.yuta.projects.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import android.widget.ToggleButton
import androidx.fragment.app.viewModels
import androidx.lifecycle.viewModelScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.yuta.app.R
import com.yuta.common.ui.BaseFragment
import com.yuta.common.util.UserUtils
import com.yuta.domain.model.ProjectDto
import com.yuta.projects.ui.adapter.ProjectsAdapter
import com.yuta.projects.ui.dialog.CreateProjectDialog
import com.yuta.projects.viewmodel.ProjectsViewModel
import kotlinx.coroutines.launch

class ProjectsFragment : BaseFragment() {

    private val createProjectButton: Button by lazy { requireView().findViewById(R.id.create_project) }
    private val managedProjectsButton: ToggleButton by lazy { requireView().findViewById(R.id.manager_button) }
    private val memberProjectsButton: ToggleButton by lazy { requireView().findViewById(R.id.member_button) }
    private val emptyText: TextView by lazy { requireView().findViewById(R.id.empty_text) }
    private val projectsAdapter: ProjectsAdapter by lazy {
        ProjectsAdapter(mutableListOf(), this) { updateLists() }
    }

    private val projectsViewModel: ProjectsViewModel by viewModels()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        return inflater.inflate(R.layout.fragment_projects, container, false).also {
            setupRecyclerView()
            setupToggleButtons()
            setupViews()
        }
    }

    override fun onStart() {
        super.onStart()
        updateLists()
    }

    private fun updateLists() {
        projectsViewModel.lastPickedButtonId?.let {
            updateLists(requireView().findViewById(it))
        }
    }

    private fun updateLists(button: View) {
        showProgress(true)
        updateProjects {
            button.performClick()
            showProgress(false)
        }
    }

    private fun updateProjects(onUpdateCallback: () -> Unit) {
        projectsViewModel.viewModelScope.launch {
            projectsViewModel.getAll(UserUtils.getUserId(requireContext())).collect {
                projectsViewModel.initializeProjects(it)
                onUpdateCallback()
            }
        }
    }

    private fun fillProjects(projectMembers: List<ProjectDto>) {
        emptyText.visibility = if (projectMembers.isEmpty()) VISIBLE else GONE
        projectsAdapter.refillList(projectMembers)
    }

    private fun onToggleButtonClick(view: View) {
        val button = view as ToggleButton
        val otherButton = if (button.id == managedProjectsButton.id) memberProjectsButton else managedProjectsButton

        if (button.id == managedProjectsButton.id) {
            fillProjects(projectsViewModel.managedProjectsMembers)
        } else {
            fillProjects(projectsViewModel.othersProjectsMembers)
        }

        setActiveButton(button)
        setInactiveButton(otherButton)
        projectsViewModel.lastPickedButtonId = button.id
    }

    private fun setupViews() {
        createProjectButton.setOnClickListener { openCreateProjectDialog() }
    }

    private fun setupRecyclerView() {
        requireView().findViewById<RecyclerView>(R.id.recyclerView).apply {
            layoutManager = LinearLayoutManager(requireContext())
            adapter = projectsAdapter
        }
    }

    private fun setupToggleButtons() {
        managedProjectsButton.setOnClickListener(this::onToggleButtonClick)
        memberProjectsButton.setOnClickListener(this::onToggleButtonClick)

        projectsViewModel.lastPickedButtonId = projectsViewModel.lastPickedButtonId ?: memberProjectsButton.id
    }

    private fun openCreateProjectDialog() = CreateProjectDialog(fragment = this) { updateLists() }.start()

    private fun setActiveButton(button: ToggleButton) {
        button.setTextAppearance(R.style.active_toggle)
        button.isChecked = true
    }

    private fun setInactiveButton(button: ToggleButton) {
        button.setTextAppearance(R.style.default_toggle)
        button.isChecked = false
    }
}
